Add plugin for Azure DNS

darshanime opened the pull request:
Signed-off-by: darshanime <deathbullet@gmail.com>

<!--
Thank you for contributing to CoreDNS!
Please provide the following information to help us make the most of your pull request:
-->

Add a plugin for Azure DNS called `azure`, like we have `route53` for AWS.

### 3. Which documentation changes (if any) need to be made?
Added a README.

### 4. Does this introduce a backward incompatible change or deprecation?
No.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add plugin fo..." ]
>+package azureDNS

this would be better named 'azure', route53 is also just that.

yongtang commented:
+1 for renaming to `azure`, also, can you fix the stickler-ci warnings?

darshanime commented:
I have changed the name, fixed stikcler-ci warnings
@miekg @yongtang, please review.

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,87 @@
+    [azure_client_id <azure_client_id>]
+    [azure_client_secret <azure_client_secret>]
+    [azure_subscription_id <azure_subscription_id>]
+    upstream
```
`upstream` is a keyword is not needed anymore. Just make an upstream.New() in the code; that get's used is there is a forward (or something else to resolve) in this coredns

darshanime replied:
gotcha, removed.

yongtang commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,139 @@
+	dnsClient := dns.NewRecordSetsClient(envSettings.Values[auth.SubscriptionID])
+	dnsClient.Authorizer, err = envSettings.GetAuthorizer()
+	if err != nil {
+		return c.Errf("\nfailed to create azure plugin:\n\n %v\n", err)
```
Is it necessary to prefix and suffix with `\n`? I think those `\n` could be removed.

yongtang commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,226 @@
+		for _, zoneName := range inputZoneNames {
+			_, err := dnsClient.ListAllByDNSZone(context.Background(), resourceGroup, zoneName, nil, "")
+			if err != nil {
+				log.Fatalf("err not nil:%s", err.Error())
```
The error should be more descriptive than `err not nil`.

yongtang commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,226 @@
+			}
+
+			if _, ok := zones[zoneName]; !ok {
+				zoneNames = append(zoneNames, fmt.Sprintf("%s.", zoneName))
```
Not very sure about `"%s."`, can you explain a little about the intention here?

darshanime replied:
cmiiw, CoreDNS adds a final trailing dot needed for the root of the DNS tree and makes the `qname` a fqdn. 
In the `zones` field of `Azure` struct, I added the same so that in `ServeDNS`, `plugin.Zones(h.zoneNames).Matches(qname)` results in a successful lookup.

yongtang replied:
I know fqdn with dot, but I am asking about:
```
if _, ok := zones[zoneName]; !ok {
```

I would expect the zoneName is pre-normalized with (or without dot), then having one lookup only?

darshanime replied:
hmm, if I understand correctly, you mean that `zoneName` would already be what it needs to be(be with or without dot `.`) in the `Corefile`? 
This won't work because azure doesn't allow creation of DNS zones with a trailing dot (throws `InvalidResourceNameTrailingCharacters`).
So, the user will always have a DNS zone like `foo.com`, we will fetch the records for `foo.com` using azure sdk but when serving via CoreDNS, it will query for a fqdn and lookup will fail if we don't make this transformation. 

let me know if I misunderstood your question though!

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,83 @@
+## Syntax
+
+~~~ txt
+azure [resource-group:dns-zone] {
```
(i'm going over this PR slowly - hence a trickly of comments)

This syntax need to be styled different, see route53 readme for instance:
`<azure_tenant_id>` -> `TENTANT_ID` (for instance). 
Also not sure if prefixing everything with `azure` helps in readability.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add plugin fo..." ]
>yongtang commented on this pull request.
>
>
>
>> +	zoneNames []string
>+	client    azure.RecordSetsClient
>+	upstream  *upstream.Upstream
>+	zMu       sync.RWMutex
>+	zones     zones
>+}
>+
>+// New validates the input dns zones and initializes the Azure struct
>+func New(ctx context.Context, dnsClient azure.RecordSetsClient, keys map[string][]string, up *upstream.Upstream) (*Azure, error) {
>+	zones := make(map[string][]*zone, len(keys))
>+	zoneNames := make([]string, 0, len(keys))
>+	for resourceGroup, inputZoneNames := range keys {
>+		for _, zoneName := range inputZoneNames {
>+			_, err := dnsClient.ListAllByDNSZone(context.Background(), resourceGroup, zoneName, nil, "")
>+			if err != nil {
>+				log.Fatalf("err not nil:%s", err.Error())
>
>The error should be more descriptive than `err not nil`.

It should also be just returned and let the caller of New decide what to do. 
Note only loop and metrics call log.Fatalf currently, all other setups errors 
are handled by just returning an error from setup()

yongtang commented:
No I mean you should normalize zoneName before hand, so that if it does not contain dot it will be added. This should happen before you perform a lookup of

_, ok = zones[zoneName]

Sent from my iPhone

On Jul 4, 2019, at 2:32 PM, Darshan Chaudhary <notifications@github.com<mailto:notifications@github.com>> wrote:


@darshanime commented on this pull request.

________________________________

In plugin/azure/azure.go<https://github.com/coredns/coredns/pull/2945#discussion_r300245739>:

> +     zones     zones
+}
+
+// New validates the input dns zones and initializes the Azure struct
+func New(ctx context.Context, dnsClient azure.RecordSetsClient, keys map[string][]string, up *upstream.Upstream) (*Azure, error) {
+       zones := make(map[string][]*zone, len(keys))
+       zoneNames := make([]string, 0, len(keys))
+       for resourceGroup, inputZoneNames := range keys {
+               for _, zoneName := range inputZoneNames {
+                       _, err := dnsClient.ListAllByDNSZone(context.Background(), resourceGroup, zoneName, nil, "")
+                       if err != nil {
+                               log.Fatalf("err not nil:%s", err.Error())
+                       }
+
+                       if _, ok := zones[zoneName]; !ok {
+                               zoneNames = append(zoneNames, fmt.Sprintf("%s.", zoneName))


hmm, if I understand correctly, you mean that zoneName would already be what it needs to be(be with or without dot .) in the Corefile?
This won't work because azure doesn't allow creation of DNS zones with a trailing dot (throws InvalidResourceNameTrailingCharacters).
So, the user will always have a DNS zone like foo.com<http://foo.com>, we will fetch the records for foo.com<http://foo.com> using azure sdk but when serving via CoreDNS, it will query for a fqdn and lookup will fail if we don't make this transformation.

let me know if I misunderstood your question though!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/coredns/coredns/pull/2945?email_source=notifications&email_token=ABU4O7AVIH3MB2PNAFB5KR3P5WKQVA5CNFSM4H4MDBRKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5PHPTQ#discussion_r300245739>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABU4O7B4PW6C6K5BFTTX3SDP5WKQVANCNFSM4H4MDBRA>.

darshanime commented:
Ah, I gotcha. Sorry for the confusion. Made the change 🔝

darshanime commented:
Hello! Can we merge this?

yongtang commented on go.mod:
```diff
@@ -67,4 +72,4 @@ require (
 	sigs.k8s.io/yaml v1.1.0 // indirect
 )
 
-replace github.com/miekg/dns v1.1.3 => github.com/miekg/dns v1.1.15
```
We want to leave 1.15 alone, as it is used to rewrite caddy's miekg/dns dependency to match the one CoreDNS depending on.

yongtang commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,223 @@
+				return nil, err
+			}
+			// Normalizing zoneName to make it fqdn
+			zoneNameFQDN := fmt.Sprintf("%s.", zoneName)
```
I think normalization might only be needed when zoneName does not ends with `.`. In other words, we could check if zoneName ends with `.`, and append only if it does not?

yongtang commented on plugin/chaos/zowners.go:
```diff
@@ -1,4 +1,4 @@
 package chaos
 
 // Owners are all GitHub handlers of all maintainers.
-var Owners = []string{"bradbeam", "chrisohaver", "dilyevsky", "ekleiner", "fastest963", "greenpau", "grobie", "inigohu", "isolus", "johnbelamaric", "miekg", "nchrisdk", "nitisht", "pmoroney", "rajansandeep", "rdrozhdzh", "rtreffer", "stp-ip", "superq", "varyoo", "yongtang"}
```
There is a `no new line at end of file` after your update. Maybe you could exclude this file from the change?

miekg commented on plugin/azure/azure.go:
```diff
@@ -45,8 +46,11 @@ func New(ctx context.Context, dnsClient azure.RecordSetsClient, keys map[string]
+			if !strings.HasSuffix(zoneNameFQDN, ".") {
+				zoneNameFQDN = fmt.Sprintf("%s.", zoneNameFQDN)
+			}
 			if _, ok := zones[zoneNameFQDN]; !ok {
```
you can just use dns.Fqdn(zonename) in this look up and delete all the lines above

miekg commented:
> Hello! Can we merge this?

We're all volunteers here, so this won't get things merged faster (in fact it might slow things down).

I also still want to review this pr

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+	zones:     returnTestZones(),
+}
+
+func returnTestZones() zones {
```
returnTestZones -> testZones

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+	return zones
+}
+
+func returnTestHandler() test.HandlerFunc {
```
-> testHandler

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+		qname := state.Name()
+		m := new(dns.Msg)
+		rcode := dns.RcodeServerFailure
+		if qname == "example.gov." {
```
example.org

darshanime replied:
this is to test for fallthrough, it should be `example.gov` no?

miekg replied:
ok, that warrents a comment then

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+			m.SetReply(r)
+			rr, err := dns.NewRR("example.gov.  300 IN  A   2.4.6.8")
+			if err != nil {
+				log.Fatalf("Failed to create Resource Record: %v", err)
```
return the err

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+	"github.com/coredns/coredns/plugin/pkg/fall"
+	"github.com/coredns/coredns/plugin/test"
+	crequest "github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
```
newline before miekg/dns

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+
+	z, ok := h.zones[zName]
```
when does this happen if we're already are authoritative for the zone? Please add a comment

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+	}, nil
+}
+
+// Run updates the zones once and starts a goroutine to do it every minute
```
dot at the end of this sentence

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+	})
+}
+
+func setup(c *caddy.Controller) error {
```
I don't see you're restarting the go-routines to update the zones. Is this handled correctly after a reload?

darshanime replied:
Hmm, I tried this with `reload` plugin and the updates seem to work as is. 
CMIIW but doesn't [`instance.Restart(corefile)`](https://github.com/coredns/coredns/blob/f8bba51f844a9ba9add993dcd579288d3af3707e/plugin/reload/reload.go#L83) take care of this? I don't see `routet53` needing to handle this either.

miekg replied:
ok

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,83 @@
+
+~~~ txt
+azure [resource-group:dns-zone] {
+    [auth_location <path to auth_location>]
```
we also usually don't have block quotes around these item, but say what the most minimal config looks like and then talk about extended config options (without the [ and ]) that are also possible.

Please take a look at route53/README.md and copy that.

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+		if result.RecordSetProperties.ARecords != nil {
+			for _, a := range *(result.RecordSetProperties.ARecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "A", *(a.Ipv4Address))
+				r, err := dns.NewRR(rfc1035)
```
This looks like a quite round about way of creating a new RR. You don't need to convert things to a string and parse the strings again. You can create a `dns.A` directly from these elements. Also removes the need to check for an error, becaue there is none

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+func updateZoneFromResourceSet(recordSet azure.RecordSetListResultPage, zName string) *file.Zone {
+	newZ := file.NewZone(zName, "")
+	for _, result := range *(recordSet.Response().Value) {
+		if result.RecordSetProperties.ARecords != nil {
```
a switch might be slightly more readable here. Also there are only 4 types a records in these zones? A/AAAA/NS and SOA, that seeems odd? If so this def. needs to be called out in the README.

darshanime replied:
There are other Record types supported too, added them 👍

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/plugin/pkg/upstream"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/dns/mgmt/dns"
```
other files have named this import "azure", please do this everywhere

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+			case "fallthrough":
+				fall.SetZonesFromArgs(c.RemainingArgs())
+			default:
+				return envSettings, keys, fall, c.Errf("unknown property '%s'", c.Val())
```
`envSettings` can be shortened to just `env`

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+type zone struct {
+	id  string
+	z   *file.Zone
+	dns string
```
`dns` is oddly named, file.Zone also has a field origin that you mean to reuse that field? If it needs to be here (I think you can use file.Zone.origin) is should be named `name` or `origin`

darshanime replied:
renamed it to `zone`

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
```
What's stopping this goroutine?

darshanime replied:
ah, missed passing it to [h.client.ListByDNSZone](https://github.com/coredns/coredns/pull/2945/files#diff-542a240e942503d1bcca9fd6f204f727R100)

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+	errc := make(chan error)
+	defer close(errc)
+	for zName, z := range h.zones {
+		go func(zName string, z []*zone) {
```
Is it important that these zones are updated in a go-routine each, or can update them in sequence (and so away with all the channel business here)

darshanime replied:
this is inspired by `route53` which uses multiple goroutines to update the zones.

miekg replied:
I think this can be just sequential

darshanime replied:
sure will do, should I make it sequential for route53 too?

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,227 @@
+	case file.NoData:
+	case file.NameError:
+		m.Rcode = dns.RcodeNameError
+	case file.Delegation:
```
are delegations supported in azure dns? Seeing only 4 record type are supported in this plugin, delegations seems like an advanced feature (although NS records are there)

darshanime replied:
Deletations are supported - 
> Azure DNS allows you to host a DNS zone and manage the DNS records for a domain in Azure. 

https://docs.microsoft.com/en-us/azure/dns/dns-domain-delegation

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/pkg/fall"
+	"github.com/coredns/coredns/plugin/test"
+	crequest "github.com/coredns/coredns/request"
```
why are renaming the request import? No other plugin does this

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+		r, _ := dns.NewRR(rr)
+		err := newZ.Insert(r)
+		if err != nil {
+			log.Fatalf("XXXXXXXXXXXXXX could not add file zone to zone: %s", err.Error())
```
xxxxxx ?

miekg replied:
also `log` in a test? Just return the error

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+		rcode := dns.RcodeServerFailure
+		if qname == "example.gov." {
+			m.SetReply(r)
+			rr, err := dns.NewRR("example.gov.  300 IN  A   2.4.6.8")
```
dns.NewRR - > test.A

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,149 @@
+	}
+}
+
+func TestAzure(t *testing.T) {
```
how about non-supported RR types?

darshanime replied:
added more tests

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	pkgFall "github.com/coredns/coredns/plugin/pkg/fall"
```
why this import rename?

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		return c.Errf("failed to create azure plugin: %v", err)
+	}
+	h, err := New(ctx, dnsClient, keys, up)
```
nit: upstream.New() can just be called here, no need to assign to `up`

miekg commented on plugin/azure/setup_test.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/caddyserver/caddy"
+)
+
+func TestSetupRoute53(t *testing.T) {
```
uhm. Azure? Or just TestSetup

miekg commented:
this is heading in the right direction, but still needs quite a bit of work

yongtang commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,83 @@
+}
+
+~~~
+Enable route53 with explicit Azure credentials:
```
`route53` -> `azure dns`?

palash25 commented:
Hi @darshanime this was supposed to be my project under the community bridge program. Since you have already worked on this I don't think I will be able to work on my project but at least go through the whole issue thread before sending a PR the next time https://github.com/coredns/coredns/issues/2820#issuecomment-499859125

darshanime commented:
Hi @palash25, 
I was not aware of that when I started working on this, apologies for that. However, since I did this indepedently (and not under the community bridge programme), the CoreDNS slot is still open and you may be able to apply with some other issue. Best of luck!

darshanime commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+				} else {
+					return envSettings, keys, fall, c.ArgErr()
+				}
+				os.Setenv("AZURE_AUTH_LOCATION", path)
```
the [`auth.GetSettingsFromFile`](https://github.com/coredns/coredns/pull/2945/files#diff-12ffff5bbde7e904a84524ed836ceaa9R92) needs the env var. 
```
func GetSettingsFromFile() (FileSettings, error) {
	s := FileSettings{}
	fileLocation := os.Getenv("AZURE_AUTH_LOCATION")
	if fileLocation == "" {
		return s, errors.New("environment variable AZURE_AUTH_LOCATION is not set")
	}
        ...
```

We can remove the feature to supply creds via a file altogether.

miekg replied:
yes, please figure something out, this is a really weird way of doing things

darshanime replied:
sure, I'll see if there is an alternative way, else will scrape the `auth_location` directive.

darshanime commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,142 @@
+
+func parseCorefile(c *caddy.Controller) (auth.EnvironmentSettings, map[string][]string, pkgFall.F, error) {
+	keyPairs := map[string]struct{}{}
+	keys := map[string][]string{}
```
`keyPairs` is a set for `resource-set:hosted-zone` entries in `Corefile`. 
`keys` stores the mappings `resource-set: [hosted-zone1, hosted-zone2]`. 

We can do away with `keyPairs` but then we will have to code to check for presence in the `[]string` slice of the corresponding `resource-set` in `keys`which is uglier.

miekg replied:
don't sound this is critical for performance. Also the naming can be improved. `keys` is very generic, cause it is in effect zones?

darshanime replied:
renamed `keys` -> `resourceGroupSet`, `keyPairs` -> `resourceGroupMapping`

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,83 @@
+*   **`auth_location`** the path to the Azure CLI authentication file. You can also provide the credentials individually (`tenant_id`, `client_id`, `client_secret`, `subscription_id`)
+
+*   `fallthrough` If zone matches and no record can be generated, pass request to the next plugin.
+    If **[ZONES...]** is omitted, then fallthrough happens for all zones for which the plugin is
```
drop the block quotes here. (This should be changed in route53 as well)

darshanime commented:
@miekg @yongtang this is ready for review again
- [x] added all remaining DNS resource records supported by Azure, added specs
- [x] updated readme
- [x] rebased

miekg commented on plugin/azure/setup_test.go:
```diff
@@ -0,0 +1,72 @@
+    foobar
+}`, true},
+		{`azure resource-set:hosted-zone {
+    tenant_id <tenant_id>
```
don't bring these <foo> in back in the tests. This also still needs to be fixed in the README.md itself.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add plugin fo..." ]
>sure will do, should I make it sequential for route53 too?

... probably? Depends on why this was original done? Also depends how much 
in sync you want to keep the codebase for these 2 plugins /cc @dilyevsky

darshanime commented:
@miekg @yongtang please take a look

- [x] addressed PR comments
- [x] rebased

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,52 @@
+## Syntax
+
+~~~ txt
+azure resource-group:dns-zone {
```
please fix this readme (again)

darshanime replied:
Sorry for the back and forth on the README. 
I am not sure about this particular case though, what changes do are you pointing at here?

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,293 @@
+	"github.com/coredns/coredns/plugin/pkg/upstream"
+	"github.com/coredns/coredns/request"
+
+	azure "github.com/Azure/azure-sdk-for-go/profiles/latest/dns/mgmt/dns"
```
is good to rename is, but doesn't this rename now clash the with package name? Maybe azuredns as an alternative?

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,293 @@
+
+type zones map[string][]*zone
+
+// Azure is the core struct of the azure plugin
```
closing dot

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,293 @@
+	zones     zones
+}
+
+// New validates the input dns zones and initializes the Azure struct
```
s/dns/DNS/
`... returns an initialized pointer to an Azure structure.`

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,293 @@
+	}, nil
+}
+
+// Run updates the zones once and starts a goroutine to do it every minute.
```
`Run updates the zone from azure.`

miekg commented on plugin/azure/azure.go:
```diff
@@ -0,0 +1,293 @@
+	return newZ
+}
+
+// ServeDNS uses the azure plugin to serve dns requests
```
`ServeDNS implements the plugin.Handler interface.`

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,180 @@
+	"github.com/miekg/dns"
+)
+
+var demoAzure Azure = Azure{
```
```suggestion
var demoAzure  = Azure{
```

miekg commented on plugin/azure/azure_test.go:
```diff
@@ -0,0 +1,180 @@
+var demoAzure Azure = Azure{
+	Next:      testHandler(),
+	Fall:      fall.Zero,
+	zoneNames: []string{"example.org", "www.example.org", "example.org.", "sample.example.org."},
```
half of these zones are not fully qualified.

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,128 @@
+	"github.com/coredns/coredns/plugin/pkg/upstream"
+
+	azure "github.com/Azure/azure-sdk-for-go/profiles/latest/dns/mgmt/dns"
+	azureCloud "github.com/Azure/go-autorest/autorest/azure"
```
azureCloud - azurerest (don't use uppercase for imports)

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,128 @@
+func setup(c *caddy.Controller) error {
+	env, keys, fall, err := parse(c)
+	if err != nil {
+		return err
```
return a `plugin.Error("azure", err)` here.

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,128 @@
+		for c.NextBlock() {
+			switch c.Val() {
+			case "subscription_id":
+				if c.NextArg() {
```
you can drop all the `else` parts here (and below) if you reverse the check
~~~
if !c.NextArg() {
   return ...
}
// set the value

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add plugin fo..." ]
>Sorry for the back and forth on the README.
>I am not sure about this particular case though, what changes do are you pointing at here?

values are CAPITALIZED, as they are in the route53 (and all other plugins)

miekg commented:
@yongtang I believe the code is ready to be merged, you?

The documentation still needs some work

darshanime commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,52 @@
+
+*   **ZONES** zones it should be authoritative for. If empty, the zones from the configuration block
+
+*   `environment` the azure environment to use. Defaults to `AzurePublicCloud`. Possible values: `AzureChinaCloud`, `AzureGermanCloud`, `AzurePublicCloud`, `AzureUSGovernmentCloud`.
```
The example mentions a minimal configuration. `environment` has a default value and need not be mentioned always. 
This documentation is here to let the user know that this is a possible configuration that can be changed.

darshanime commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,52 @@
+
+~~~ txt
+azure RESOURCE-GROUP:DNS-ZONE {
+    tenant_id TENANT_ID
```
`tenant_id`, `subscription_id` etc is the jargon used in the azure community. Should I drop the `_id`s?

darshanime replied:
chaning the key to `tenant` but README value as `TENANT_ID` will keep things clear.

darshanime commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,52 @@
+*   **`dns-zone`** the dns zone that contains the resource record sets to be
+    accessed.
+
+*   `fallthrough` If zone matches and no record can be generated, pass request to the next plugin.
```
the example above is minimal, should I include all possible configurations? How do I indicate it's optional without using the `[]` braces

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+## Syntax
+
+~~~ txt
+azure RESOURCE_GROUP:ZONE {
```
```suggestion
azure RESOURCE_GROUP:ZONE... {
    tenant TENANT
    client CLIENT SECRET
    subscription SUBSCRIPTION_ID
}
~~~

* **RESOURCE_GROUP:ZONE** The resource group to which the hosted zones belong on Azure and the 
  **ZONE** where the *azure* plugin should be authoritative for. 

* `client` defines the **CLIENT** and **SECRET** used to connect to Azure.
```

miekg commented on plugin/azure/setup.go:
```diff
@@ -0,0 +1,123 @@
+					return env, resourceGroupMapping, fall, c.ArgErr()
+				}
+				env.Values[auth.ClientSecret] = c.Val()
+			case "environment":
```
what does environment does that secret doesn't do? Is set the secret and then uses that value to get the env from azure?
Is that how azure works?

darshanime replied:
`environment` is used to talk to non public clouds, like  `AzureChinaCloud`, `AzureGermanCloud`, `AzurePublicCloud`, `AzureUSGovernmentCloud` etc. 
By default, the plugin talks to `AzurePublicCloud`

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+
+## Description
+
+The azure plugin is useful for serving zones from resource record
```
```suggestion
The *azure* plugin is useful for serving zones from resource record
```

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+## Description
+
+The azure plugin is useful for serving zones from resource record
+sets in Microsoft Azure DNS. The `azure` plugin can be used when coredns is deployed on Azure or elsewhere.
```
```suggestion
from Microsoft Azure DNS. 
```

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+
+The azure plugin is useful for serving zones from resource record
+sets in Microsoft Azure DNS. The `azure` plugin can be used when coredns is deployed on Azure or elsewhere.
+This plugin supports all the DNS records supported by Azure, viz. A, AAAA, CAA, CNAME, MX, NS, PTR, SOA, SRV, and TXT record types.
```
```suggestion
The *azure* plugins supports all the DNS records supported by Azure, viz. A, AAAA, CAA, CNAME, MX, NS, PTR, SOA, SRV, and TXT record types. For other types the result with be NOTIMPL.
```

miekg replied:
what is the result for a query for a type that is not supported?

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+
+## Examples
+
+Enable `azure` with Azure credentials and fallthrough:
```
```suggestion
Enable the *azure* plugin with Azure credentials and fallthrough:
```

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+
+~~~ txt
+. {
+    azure resource_group_foo:foo.com {
```
this makes azure handle `foo.com`, the the fallthrough has `example.org`. example.org names will never be handled by the *azure* plugin so this example doesn't work.

Also I would just list a working minimal example first and then expand on that.

miekg commented on plugin/azure/README.md:
```diff
@@ -0,0 +1,50 @@
+      fallthrough example.org.
+    }
+}
+~~~
```
Are there links to Azure that we can put in a `## Also See` section for folks interested in Azure DNS?

miekg commented:
I think this is good to go - I might have some nits, but I'll just send a follow PR.

Expect an invite for becoming a maintainer soon and can you ping me your email addr (miek@miek.nl) so I can add you to the maintainer lists? Thanks!

You only need to fix the conflicts - then we're good to go.
